File: ISP/java/ISPIntact.java
interface Shape2D {
    int area();
}

interface Shape3D extends Shape2D {
    int volume();
}

class Square implements Shape2D {
    private int side;

    public Square(int side) { 
        this.side = side;
    }

    public int area() { 
        return side * side; 
    }
}

class Rectangle implements Shape2D {
    private int length, breadth;

    public Rectangle(int length, int breadth) {
        this.length = length;
        this.breadth = breadth;
    }

    public int area() { 
        return length * breadth; 
    }
}

class Cube implements Shape3D {
    private int side;

    public Cube(int side) { 
        this.side = side; 
    }

    public int area() { 
        return side * side * 6; 
    }

    public int volume() { 
        return side * side * side; 
    }
}

public class ISPIntact {
    public static void main(String[] args) {
        Shape2D square = new Square(6);
        Shape2D rectangle = new Rectangle(3, 6);
        Shape3D cube = new Cube(6);

        System.out.println("Square Area: " + square.area());
        System.out.println("Rectangle Area: " + rectangle.area());
        System.out.println("Cube Area: " + cube.area());
        System.out.println("Cube Volume: " + cube.volume());
    }
}

File: ISP/isp_intact.cpp
#include <iostream>
using namespace std;

// Fat shape interface is now segragated into two different interfaces
class Shape2D {
public:
    virtual int area() = 0;    
};

class Shape3D : public Shape2D {
public:
    virtual int volume() = 0;
};

// Square is a 2D shape hence it only implements the area
class Square : public Shape2D {
private:
    int side;

public:
    Square(int side) {
        this->side = side;
    }

    int area() override {
        return side * side;
    }
};

// Rectangle now only implements area
class Rectangle : public Shape2D {
private:
    int length, breadth;

public:
    Rectangle(int length, int breadth) {
        this->length = length;
        this->breadth = breadth;
    }

    int area() override {
        return length * breadth;
    }
};

// Cube implements both, area and volume since it is a 3d shape
class Cube : public Shape3D {
private:
    int side;

public:
    Cube(int side) {
        this->side = side;
    }

    int area() override {
        return side * side * 6;
    };

    int volume() override {
        return side * side * side;
    }

};

int main() {
    Shape2D* square = new Square(6);
    Shape2D* rectangle = new Rectangle(3, 6);
    Shape3D* cube = new Cube(6);

    cout << "Square Area: " << square->area() << endl;
    cout << "Rectangle Area: " << rectangle->area() << endl;
    cout << "Cube Area: " << cube->area() << endl;
    cout << "Cube Volume: " << cube->volume() << endl;

    delete square;
    delete rectangle;
    delete cube;

    return 0;
}